### PR TITLE
 Fixed a bug in hardClipCigar function that caused incorrect cigar calculation

### LINKED ISSUE #6139: ReadClipper::hardClipByReadCoordinates incorrectly clips cigars in the presence of indels
In the presence of insertions or deletions in the cigar, hardClipByReadCoordinates gets confused regarding the amount of bases that were clipped, thereby creating an incorrect cigar. The following snippet, placed in ReadClipperUnitText.java can serve as a test.

```java
@DataProvider
    Object[][] cigarsToClipData(){
        return new Object[][]{
                new Object[]{"20M", 6},
                new Object[]{"3M2I20M", 4},
                new Object[]{"10I20M", 0},
                new Object[]{"3M2D20M", 6},
        };
    }

        public void testHardClipSoftClippedBasesClipsTheCorrectAmount(final String cigarString, final int alignmentOffset) {
        final int start = 100;
        final GATKRead originalRead = ArtificialReadUtils.createArtificialRead(TextCigarCodec.decode(cigarString));
        BaseUtils.fillWithRandomBases(originalRead.getBasesNoCopy(),0,originalRead.getLength());
        originalRead.setPosition(originalRead.getContig(), start);

        final GATKRead clippedRead = ReadClipper.hardClipByReadCoordinates(originalRead, 0, 5);
        Assert.assertEquals(
                clippedRead.getCigar().getReadLength() + AlignmentUtils.getNumHardClippedBases(clippedRead),
                originalRead.getCigar().getReadLength(), " Clipped cigar: " + clippedRead.getCigar());
        Assert.assertEquals(clippedRead.getStart(),start + alignmentOffset, " Clipped cigar: " + clippedRead.getCigar());
        Assert.assertEquals(clippedRead.getBasesNoCopy(),Arrays.copyOfRange(originalRead.getBases(),5,originalRead.getBasesNoCopy().length));
        Assert.assertEquals(clippedRead.getBaseQualitiesNoCopy(),Arrays.copyOfRange(originalRead.getBases(),5,originalRead.getBaseQualitiesNoCopy().length));
    }

```

### PR DESCRIPTION
**Pull request by yfarjoun**
Problem was ocurring in the presence of insertions and deletions. fixes #6139

1. Changed ReadClipper unit tests:
  - The tests in many cases assumed that the unclipped alignment locations do not    change when the read is clipped. This is not true: for example if start for cigar
   1M1I3M is 100, the unclipped start for 2H3M is 99. All assertUnclipped calls were
   removed
   - Alignment now check that the read length remains to be consistent with the CIGAR, that the aligned bases span are consistent with the CIGAR and that the number of clipped bases from the read is consistent with the requested clipping
2. Hard clipping in ClippingOps was buggy, thus we introduced new tests for it.
3. Text was refactored for readability
4. Clipping in ClippingOps did not treat insertions and deletions in the clipped parts of the CIGAR correctly. This was fixed
5. Alignment re-calculation after clipping did not work correctly if the initial CIGAR contained insertions and deletions
6. Hard clipping applied to the hard clipped read did not behave correctly

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 1]
```diff
@@ -548,7 +545,8 @@ private CigarShift cleanHardClippedCigar(final Cigar cigar) {
                 } else if (!readHasStarted && cigarElement.getOperator() == CigarOperator.HARD_CLIP) {
                     totalHardClip += cigarElement.getLength();
                 } else if (!readHasStarted && cigarElement.getOperator() == CigarOperator.DELETION) {
-                    totalHardClip += cigarElement.getLength();
+                    //totalHardClip += cigarElement.getLength();
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: It seems like you are putting in an effect-less line of code to document that the total is not incremented.  Is that because there are a bunch of `else if`s and you don't want to give the impression that it's always incremented?  How about this:
```
if (!readHasStarted && (cigarElement.getOperator() == CigarOperator.HARD_CLIP || cigarElement.getOperator() == CigarOperator.CigarOperator.SKIPPED_REGION) {
  totalHardClip += cigarElement.getLength();
}
readHasStarted |= cigarElement.getOperator() != CigarOperator.DELETION &&
                        cigarElement.getOperator() != CigarOperator.SKIPPED_REGION &&
                        cigarElement.getOperator() != CigarOperator.HARD_CLIP;
```

Also, in the (unmodified) code below it seems that the `if (ReadHasStartedBlock)` always results in `addedHardClips == true`.  If you agree, could you move the `addedHardCips = true` statements to the end of that block?

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 2]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+        while (i < elements.size() && elements.get(i).getOperator() == CigarOperator.SOFT_CLIP) {
             size += elements.get(i).getLength();
             i++;
         }
 
         return size;
     }
 
+    /**
+    * Calculates shift of alignment int the newCigar relative to the oldCigar due to
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: `int` -> `in`

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 3]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+     * hard/soft clipping under assumption that the
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: This line ends too early.

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 4]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+     * @param newCigar new cigar (with hard/soft clipping)
+     * @return int the offset (from 0 between the alignment start on the old and on the new cigar)
+    */
     private int calculateAlignmentStartShift(final Cigar oldCigar, final Cigar newCigar) {
         final int newShift = calcHardSoftOffset(newCigar);
         final int oldShift = calcHardSoftOffset(oldCigar);
         return newShift - oldShift;
     }
 
+    /**
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Extra `/**` here

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 5]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+    /* Calculates how much the alignment should be shifted when hard/soft clipping is applied
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: This should be `/**`

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 6]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+     * @return int - the offset between the alignment starts in the oldCigar and in
+     * the cigar after applying clipping
+     */
+    private int calculateAlignmentStartShift(final Cigar oldCigar, final int clipping) {
+
+        final int newReadBasesClipped = clipping;
+        int readBasesClipped = 0; // The number of read bases consumed on the new cigar before reference bases are consumed
+        int refBasesClipped = 0; // A measure of the reference offset between the oldCigar and the clippedCigar
+
+        boolean final_element = false;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: camel case

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 7]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+
+            // needed only if the clipping ended on N or D
+            if (final_element) {
+                if (!e.getOperator().consumesReadBases() && e.getOperator().consumesReferenceBases()) {
+                    refBasesClipped += curRefLength;
+                }
+                break;
+            }
+
+            if (!e.getOperator().consumesReadBases()) {
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Can you absorb this into the above as
```
final int curReadLength = e.getOperator().consumesReadBases() ? e.getLength() : 0;
```

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 8]
```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
...
+
+            if (!e.getOperator().consumesReferenceBases()) {
+                curRefLength = 0;
+            }
+
+            readBasesClipped += curReadLength;
+            refBasesClipped += curRefLength;
+
+            if (readBasesClipped > newReadBasesClipped || truncated) {
+                break;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: The control flow is tricky to follow here because there are two fairly distant `breaks` as well as a `final_element = true` that then triggers a `break` partway through the next iteration.  Can this be made clearer?

File: src/main/java/org/broadinstitute/hellbender/utils/read/ArtificialReadUtils.java [hunk 9]
```diff
@@ -438,16 +440,16 @@ public static SAMRecord createArtificialSAMRecord(final SAMFileHeader header, fi
...
     }
 
     public static SAMRecord createArtificialSAMRecord(final SAMFileHeader header, final Cigar cigar, final String name) {
-        int length = cigar.getReadLength();
-        byte base = 'A';
-        byte qual = 30;
-        byte [] bases = Utils.dupBytes(base, length);
-        byte [] quals = Utils.dupBytes(qual, length);
+        final int length = cigar.getReadLength();
+        final Random random = new Random(TestUtil.RANDOM_SEED);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/read/ArtificialReadUtils.java]: What about `Utils.getRandomGenerator()` or some such `Utils` method?

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java [hunk 10]
```diff
@@ -87,19 +97,24 @@ public void testHardClipByReferenceCoordinates() {
         for (Cigar cigar : cigarList) {
             GATKRead read = ReadClipperTestUtils.makeReadFromCigar(cigar);
             int start = getSoftStart(read);
+            int aln_start = read.getStart();
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java]: camel case

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java [hunk 11]
```diff
@@ -323,6 +339,76 @@ private void assertUnclippedLimits(final GATKRead original, final GATKRead clipp
...
+     * Asserts that number of clipped bases from the read is consistent with the requested clipping
+     *
+     * @param original original read
+     * @param clipped clipped read
+     * @param clipping number of clipped bases requested
+     * */
+    private void assertReadClippingConsistent(final GATKRead original,
+                                              final GATKRead clipped,
+                                              final int clipping) {
+        int clip_diff = original.getLength() - clipped.getLength();
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java]: camel case

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java [hunk 12]
```diff
@@ -323,6 +339,76 @@ private void assertUnclippedLimits(final GATKRead original, final GATKRead clipp
...
+                                             final int clipped_left,
+                                             final int clipped_right){
+
+        if (original.getLengthOnReference() - clipped_left - clipped_right <= 0 ){
+            if (clipped.getLengthOnReference() == 0 )
+                return;
+            Assert.assertTrue(clipped.isUnmapped());
+            return;
+        }
+        int delta_left = clipped.getStart() - original.getStart();
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java]: camels

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java [hunk 13]
```diff
@@ -323,6 +339,76 @@ private void assertUnclippedLimits(final GATKRead original, final GATKRead clipp
...
+        Assert.assertEquals(delta_left, clipped_left);
+        int original_end = original.getEnd() ;
+        int clipped_end = clipped.getEnd() ;
+        int delta_right = original_end - clipped_end;
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ReadClipperUnitTest.java]: camels

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 14]
```diff
@@ -123,18 +123,18 @@ private GATKRead applySoftClipBases(final GATKRead readCopied, final boolean run
...
         final byte[] newQuals = readCopied.getBaseQualities(); //this makes a copy so we can modify in place
-        overwriteFromStartToStop(newQuals, (byte)0);
+        overwriteFromStartToStop(newQuals, (byte) 0);
         readCopied.setBaseQualities(newQuals);
     }
 
     private void applyWriteNs(final GATKRead readCopied) {
         final byte[] newBases = readCopied.getBases();       //this makes a copy so we can modify in place
-        overwriteFromStartToStop(newBases, (byte)'N');
+        overwriteFromStartToStop(newBases, (byte) 'N');
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: I added a space, and you want me to fix the old code!?? :-)

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 15]
```diff
@@ -258,9 +259,7 @@ private Cigar softClip(final Cigar __cigar, final int __startClipEnd, final int
 
             final Cigar newCigar = new Cigar();
             newCigar.add(new CigarElement(cigarLength, CigarOperator.SOFT_CLIP));
-            if (runAsserts) {
-                assert newCigar.isValid(null, -1) == null;
-            }
+            assert !runAsserts || newCigar.isValid(null, -1) == null;
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: reverted to keep the history cleaner

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java (hunk 8 above)
Comment (yfarjoun) [reply src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: yes...it's tricky....not sure how to make it less hard to follow.

Comment (yfarjoun) [reply src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: in resolving this I found a bug! thanks....will resubmit soon.

**Comment by yfarjoun** [comment]
builds are now failing due to the changes in the artifical reads generator...I'll make them more flexible so that I can still get more random reads without breaking the other tests....

**Comment by yfarjoun** [comment]
back to you @davidbenjamin The tests are also passing this time!

**Comment by yfarjoun** [comment]
Thanks for the review @davidbenjamin 🎉
